src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -5042,6 +5042,247 @@ class StubGenerator: public StubCodeGenerator {
     return (address) start;
   }
 
+  void generate_updateBytesAdler32_accum(Register buff, VectorRegister vzero, VectorRegister vbytes,
+    VectorRegister vs1acc, VectorRegister vs2acc, VectorRegister vtable) {
+    // Below is a vectorized implementation of updating s1 and s2 for 16 bytes.
+    // We use b1, b2, ..., b16 to denote the 16 bytes loaded in each iteration.
+    // In non-vectorized code, we update s1 and s2 as:
+    //   s1 <- s1 + b1
+    //   s2 <- s2 + s1
+    //   s1 <- s1 + b2
+    //   s2 <- s2 + b1
+    //   ...
+    //   s1 <- s1 + b16
+    //   s2 <- s2 + s1
+    // Putting above assignments together, we have:
+    //   s1_new = s1 + b1 + b2 + ... + b16
+    //   s2_new = s2 + (s1 + b1) + (s1 + b1 + b2) + ... + (s1 + b1 + b2 + ... + b16)
+    //          = s2 + s1 * 16 + (b1 * 16 + b2 * 15 + ... + b16 * 1)
+    //          = s2 + s1 * 16 + (b1, b2, ... b16) dot (16, 15, ... 1)
+    //
+    // To add more acceleration, it is possible to postpone final summation
+    // until all unrolled calculations are done.
+
+    // Load data
+    __ vle8_v(vbytes, buff);
+    __ addi(buff, buff, 16);
+
+    // vs1acc = b1 + b2 + b3 + ... + b16
+    __ vwredsumu_vs(vs1acc, vbytes, vzero);
+
+    // vs2acc = { (b1 * 16), (b2 * 15), (b3 * 14), ..., (b8 * 9) }
+    // vs2acc->successor() = { (b9 * 8), (b10 * 7), (b11 * 6), ..., (b16 * 1) }
+    __ vwmulu_vv(vs2acc, vtable, vbytes); // vs2acc->successor() now contains the second part of multiplication
+    // The only thing that remains is to sum up the remembered result
+  }
+
+  void generate_updateBytesAdler32_unroll(Register buff, Register s1, Register s2, int unroll_factor,
+    VectorRegister vtable, VectorRegister vbytes, VectorRegister vzero, VectorRegister *unroll_regs,
+    Register temp0, Register temp1, Register temp2, VectorRegister vtemp1, VectorRegister vtemp2) {
+
+    assert(unroll_factor <= 8, "Not enough vector registers in unroll_regs");
+    // Below is partial loop unrolling for updateBytesAdler32:
+    // First, the unroll*16 bytes are processed, the results are in
+    // v4, v5, v6, ..., v25, v26, v27
+    // Second, the final summation for unrolled part of the loop should be performed
+
+    __ vsetivli(temp0, 16, Assembler::e8, Assembler::m1);
+    for (int i = 0; i < unroll_factor; i++)
```

Comment:
Does it make sense to limit the vector lenth to 16 bytes and do loop unrolling here? I think the aarch64 version of `generate_updateBytesAdler32_accum` has this constraint because they use NEON which only has 128-bit vector registers. But for RVV, we can combine several vector registers into register group (LMUL greater than 1).

Reply:
Hi! Thanks for pointing it out! Sorry for such a late reply. I made some changes with vector register grouping, using LMUL = 4 mode, as this size is maximum possible with current calculating algorithm. I listed updated results below. Can you please take another look?